### PR TITLE
fix iteration and bot cleanup

### PR DESCRIPTION
**Pull request by Vigerus**

### CODE REVIEW

File: playerbot/PlayerbotMgr.cpp [hunk 1]
```diff
@@ -47,25 +63,31 @@ void PlayerbotHolder::UpdateSessions(uint32 elapsed)
...
+   auto it = playerBots.begin();
+
+   while (it != playerBots.end())
+   {
+      if (it->second == nullptr)
+      {
+         it = playerBots.erase(it);
+      }
+
+      ++it;
```
Comment (davidonete) [inline playerbot/PlayerbotMgr.cpp:82]: I believe ++it should be when not deleting

File: playerbot/PlayerbotMgr.cpp [hunk 2]
```diff
@@ -47,25 +63,31 @@ void PlayerbotHolder::UpdateSessions(uint32 elapsed)
...
-        if (itr == GetPlayerBotsEnd()) break;
-        Player* bot= itr->second;
-        if (!bot->GetPlayerbotAI()->isRealPlayer())
-            LogoutPlayerBot(bot->GetObjectGuid().GetRawValue());
-    }
-    */    
-    PlayerBotMap bots = playerBots;
-
-    for (auto& itr : bots)
+    for (auto& itr : playerBots)
```
Comment (davidonete) [inline playerbot/PlayerbotMgr.cpp:88]: Shouldn't you be using the method you created earlier to iterate between all bots?

File: playerbot/PlayerbotMgr.cpp [hunk 3]
```diff
@@ -1114,9 +1153,12 @@ void PlayerbotMgr::HandleCommand(uint32 type, const std::string& text, uint32 la
         bot->GetPlayerbotAI()->HandleCommand(type, text, *master, lang);
     }
 
-    for (PlayerBotMap::const_iterator it = sRandomPlayerbotMgr.GetPlayerBotsBegin(); it != sRandomPlayerbotMgr.GetPlayerBotsEnd(); ++it)
+    for (auto& itr : playerBots)
```
Comment (davidonete) [inline playerbot/PlayerbotMgr.cpp:1157]: This is not the same playerBots map, this one lives in RandomPlayerbotMgr

File: playerbot/PlayerbotMgr.cpp [hunk 4]
```diff
@@ -1133,18 +1175,18 @@ void PlayerbotMgr::HandleCommand(uint32 type, const std::string& text, uint32 la
...
+   {
+        Player* bot = itr.second;
         if (!bot)
             continue;
 
         bot->GetPlayerbotAI()->HandleMasterIncomingPacket(packet);
     }
 
-    for (PlayerBotMap::const_iterator it = sRandomPlayerbotMgr.GetPlayerBotsBegin(); it != sRandomPlayerbotMgr.GetPlayerBotsEnd(); ++it)
+    for (auto& itr : playerBots)
```
Comment (davidonete) [inline playerbot/PlayerbotMgr.cpp:1188]: This is not the same playerBots map, this one lives in RandomPlayerbotMgr

File: playerbot/PlayerbotMgr.cpp [hunk 5]
```diff
@@ -1182,9 +1224,9 @@ void PlayerbotMgr::HandleMasterOutgoingPacket(const WorldPacket& packet)
         bot->GetPlayerbotAI()->HandleMasterOutgoingPacket(packet);
     }
 
-    for (PlayerBotMap::const_iterator it = sRandomPlayerbotMgr.GetPlayerBotsBegin(); it != sRandomPlayerbotMgr.GetPlayerBotsEnd(); ++it)
+    for (auto& itr : playerBots)
```
Comment (davidonete) [inline playerbot/PlayerbotMgr.cpp:1228]: This is not the same playerBots map, this one lives in RandomPlayerbotMgr

File: playerbot/PlayerbotMgr.h [hunk 6]
```diff
@@ -26,15 +26,15 @@ class PlayerbotHolder : public PlayerbotAIBase
...
 
     virtual void UpdateAIInternal(uint32 elapsed, bool minimal = false);
     void UpdateSessions(uint32 elapsed);
 
+    void LoopThroughBots(std::function<void(Player*)> fct) const;
+
     void LogoutAllBots();
     void OnBotLogin(Player* bot);
-    void MovePlayerBot(uint32 guid, PlayerbotHolder* newHolder) { auto botptr = playerBots.find(guid); if (botptr == playerBots.end()) return;  newHolder->OnBotLogin(botptr->second); playerBots.erase(guid); }
+    void MovePlayerBot(uint32 guid, PlayerbotHolder* newHolder) { auto botptr = playerBots.find(guid); if (botptr == playerBots.end() || botptr->second == nullptr) return;  newHolder->OnBotLogin(botptr->second); playerBots[guid] = nullptr; }
```
Comment (davidonete) [inline playerbot/PlayerbotMgr.h:37]: Can you move this to PlayerbotMgr.cpp please?

**Review by davidonete — CHANGES_REQUESTED** [review]

File: playerbot/PlayerbotMgr.cpp [hunk 7]
```diff
@@ -97,9 +126,13 @@ void PlayerbotMgr::CancelLogout()
         }
     }
 
-    for (PlayerBotMap::const_iterator it = sRandomPlayerbotMgr.GetPlayerBotsBegin(); it != sRandomPlayerbotMgr.GetPlayerBotsEnd(); ++it)
+    for (auto& itr : playerBots)
```
Comment (davidonete) [inline playerbot/PlayerbotMgr.cpp:129]: This is not the same playerBots map, this one lives in RandomPlayerbotMgr

**Review by davidonete — CHANGES_REQUESTED** [review]

File: playerbot/PlayerbotMgr.cpp (hunk 1 above)
Comment (Vigerus) [reply playerbot/PlayerbotMgr.cpp:82]: yeah, there's continue inside, wrong cherry pick

File: playerbot/PlayerbotMgr.cpp (hunk 2 above)
Comment (Vigerus) [reply playerbot/PlayerbotMgr.cpp:88]: the idea for the exposed method is not for internal use

**Review by davidonete — APPROVED** [review]

File: playerbot/PlayerbotMgr.cpp [hunk 8]
```diff
@@ -9,30 +8,54 @@
...
 }
 
+void PlayerbotHolder::ForEachPlayerbot(std::function<void(Player*)> callback) const
+{
+    for (auto& itr : playerBots)
+    {
+        Player* bot = itr.second;
+        if (bot)
+		{
+	    	callback(bot);
```
Comment (Exxenoz) [inline playerbot/PlayerbotMgr.cpp:33]: Tabs and whitespaces mixed. 😓

Comment (davidonete) [reply playerbot/PlayerbotMgr.cpp:33]: I edited it via github, seems like it doesn't work very well

**Pull request merged**